### PR TITLE
Fix name clashes in FunctionEmitter.

### PR DESCRIPTION
**Pull request by sjrd**
~~Based on #3814. Only the last 3 commits belong to this PR.~~

### CODE REVIEW

**Comment by gzm0** [comment]
I've been trying to wrap my head around of how feasible it would be to always translate record identifiers fully. I find it really strange that we introduce intermediate record variables in JS trees that are not variables that will actually appear in the final program.

I realize that this will require a more advanced form of `Lhs.Assign`, but I'm struggling to think of other complications. @sjrd do you have any intuition about this? I feel it could simplify understanding of the code a lot if record selects are always translated fully.

**Comment by sjrd** [comment]
I tried for a while to do as you suggest, but I hit difficulties. It's difficult to explain, but I run into trouble:
* trying to adapt `doVarDef` and `doAssign` to support this change,
* trying to make `unnest` do something sensible when it needs to extract a temporary var with a record type
* `extractLet` becomes more complicated because it needs to handle records separately from non-records

Basically, I haven't found a solution quickly enough for this, and I don't think the problem is real enough for me to spend more time on it now. I suggest we revisit later (post 1.0.0).

**Comment by sjrd** [comment]
Thinking more about this, I don't understand what your concern is. It seems perfectly reasonable to me to manipulate `js.Ident` for partially transformed records, because they're records themselves. Since we can have temporary variables that are in the JS namespace, and that we can have records as temporary variables, it follows that we can have records in the JS namespace. The fact that we need to combine them further with fields before emitting them to the final JS code doesn't bother me.

**Comment by gzm0** [comment]
Given that records are a pure IR concept and do not exist in JS, I find it strange that we can have a temporary variable in the JS namespace that can hold a record (which it will not eventually of course, but during the transformation, conceptually it can).

Another angle to look at it: Hypothetically, if you would have to give that temporary variable a (JS) type what would it be?

**Comment by sjrd** [comment]
I can't give it an actual JS type, but I can imagine a phantom type that is also a record, and is temporarily used to reason about it.

File: linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/JSGen.scala [hunk 1]
```diff
@@ -623,7 +623,12 @@ private[emitter] final class JSGen(val semantics: Semantics,
     VarRef(envFieldIdent(field))
 
   def envFieldIdent(field: String)(implicit pos: Position): Ident =
-    Ident(avoidClashWithGlobalRef("$" + field))
+    envFieldIdent(field, None)
+
+  def envFieldIdent(field: String, origName: Option[String])(
```
[the remark this answers is not in the document]
Comment (sjrd) [reply linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/JSGen.scala]: Yes it's basically for any emitter-generated identifier in the variable scope. It's not the first time we do that -- we have precedent with the `$cc_...` variables for class captures -- but now it's become painfully obvious.

**Pull request merged**